packages/modules/analytics/src/collectors/BaseCollector.ts

```diff
@@ -0,0 +1,61 @@
+import type { AnyValue, AnyValueMap, AnalyticsEvent } from '../types.js';
+import type { IAnalyticsCollector } from './AnalyticsCollector.interface.js';
+
+import { from, map, type ObservableInput, Subject, type Observer, type Unsubscribable } from 'rxjs';
+import { z } from 'zod';
+
+/**
+ * Function to create zod schema to be used when collecting
+ */
+export const createSchema = <TValue = AnyValue, TAttr = AnyValueMap>(
+  value: z.ZodSchema<TValue>,
+  attributes: z.ZodSchema<TAttr>,
+) => {
+  return z.object({
+    name: z.string(),
+    value: value,
+    attributes: attributes.optional(),
+  });
+};
+
+export type CollectorSchema<TValue = AnyValue, TAttr = AnyValueMap> = ReturnType<
+  typeof createSchema<TValue, TAttr>
+>;
+
+export abstract class BaseCollector<
```

Comment:
what is a base collector and what shared functionality does it give to children?